Source/Core/DiscIO/VolumeVerifier.cpp

```diff
@@ -39,6 +49,291 @@
+      m_result = {Status::Error, Common::GetStringT("Wii data is not public yet")};
+      return {};
+
+    case DownloadStatus::Fail:
+    default:
+      m_result = {Status::Error, Common::GetStringT("Failed to connect to Redump.org")};
+      return {};
+    }
+  });
+}
+
+static std::string GetPathForSystem(const std::string& system)
+{
+  return File::GetUserPath(D_REDUMPCACHE_IDX) + DIR_SEP + system + ".zip";
+}
+
+RedumpVerifier::DownloadStatus RedumpVerifier::DownloadDatfile(const std::string& system,
+                                                               DownloadStatus old_status)
+{
+  if (old_status == DownloadStatus::Success || old_status == DownloadStatus::SystemNotAvailable)
+    return old_status;
+
+  Common::HttpRequest request;
+
+  const std::optional<std::vector<u8>> result =
+      request.Get("http://redump.org/datfile/" + system + "/serial,version",
+                  {{"User-Agent", Common::scm_rev_str}});
+
+  const std::string output_path = GetPathForSystem(system);
+
+  if (!result)
+  {
+    return File::Exists(output_path) ? DownloadStatus::FailButOldCacheAvailable :
+                                       DownloadStatus::Fail;
+  }
+
+  if (result->size() > 1 && (*result)[0] == '<' && (*result)[1] == '!')
+  {
+    // This is an HTML page, not a zip file like we want
+
+    if (File::Exists(output_path))
+      return DownloadStatus::FailButOldCacheAvailable;
+
+    const std::string system_not_available_message = "System \"" + system + "\" doesn't exist.";
+    const bool system_not_available_match =
+        result->end() != std::search(result->begin(), result->end(),
+                                     system_not_available_message.begin(),
+                                     system_not_available_message.end());
+    return system_not_available_match ? DownloadStatus::SystemNotAvailable : DownloadStatus::Fail;
+  }
+
+  File::IOFile(output_path, "wb").WriteBytes(result->data(), result->size());
+  return DownloadStatus::Success;
+}
+
+std::vector<u8> RedumpVerifier::ReadDatfile(const std::string& system)
+{
+  unzFile file = unzOpen(GetPathForSystem(system).c_str());
+  if (!file)
+    return {};
+
+  Common::ScopeGuard file_guard{[&] { unzClose(file); }};
+
+  // Check that the zip file contains exactly one file
+  if (unzGoToFirstFile(file) != UNZ_OK)
+    return {};
+  if (unzGoToNextFile(file) != UNZ_END_OF_LIST_OF_FILE)
+    return {};
+
+  // Read the file
+  if (unzGoToFirstFile(file) != UNZ_OK)
+    return {};
+  unz_file_info file_info;
+  unzGetCurrentFileInfo(file, &file_info, nullptr, 0, nullptr, 0, nullptr, 0);
+  std::vector<u8> data(file_info.uncompressed_size);
+  if (!Common::ReadFileFromZip(file, &data))
+    return {};
+
+  return data;
+}
+
+static u8 ParseHexDigit(char c)
+{
+  if (c < '0')
+    return 0;  // Error
+
+  if (c >= 'a')
+    c -= 'a' - 'A';
+  if (c >= 'A')
+    c -= 'A' - ('9' + 1);
+  c -= '0';
+
+  if (c >= 0x10)
+    return 0;  // Error
+
+  return c;
+}
```

Comment:
It feels a bit complicated to have this, but [according to StackOverflow](https://stackoverflow.com/q/17261798) there isn't really a better way.